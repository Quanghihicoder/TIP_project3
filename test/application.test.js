// Test AWS infrastructure

const cdk = require('aws-cdk-lib');
const { Template } = require('aws-cdk-lib/assertions');
const { ApplicationStack } = require('../lib/application-stack');

const Sydney = {
    account: "058264550947",
    region: "ap-southeast-2",
};

const app = new cdk.App();
const stack = new ApplicationStack(app, 'ProdApplicationStack', { env: Sydney, stackName: "ProdApplicationStack", stage: 'prod' });
const template = Template.fromStack(stack);

// ======================================================== Unit/Assertions Test =====================================================

// Test for lambda.Code.fromCfnParameters()
test('Lambda Code is created from CloudFormation parameters', () => {
    //template.resourceCountIs('AWS::CloudFormation::CustomResource', 1)
})

// Test for ssm.StringParameter.fromStringParameterName()
test('SSM Parameter for Web Crawler Bucket Location is referenced', () => {
    //template.hasResourceProperties('AWS::SSM::Parameter', {
    //    Name: 'webcrawler-bucket-location'
    //})
})

test('S3 bucket is created', () => {
    //template.hasResourceProperties('AWS::S3::Bucket', {
    //   BucketName: 'webcrawler-assets-bucket-location'
    //})
})

test('Web Crawler has been created', () => {
    template.hasResource("AWS::Lambda::Function", "");

    template.hasResourceProperties('AWS::Lambda::Function', {
        Handler: "webcrawler.handler"
    })
})

test('IAM Role for Web Crawler is created', () => {
    template.hasResource('AWS::IAM::Role', "");
    template.hasResourceProperties('AWS::IAM::Role',{
        Description: 'Web Crawler IAM Role-ap-southeast-2-prod',
        AssumeRolePolicyDocument: {
            Statement: [{
                    Effect: 'Allow',
                    Principal: {
                        Service: 'lambda.amazonaws.com'
                    },
                    Action: 'sts:AssumeRole'}]}
    })
})

test('Web Crawler runs every 5 minutes', () => {
    template.hasResourceProperties('AWS::Events::Rule', {
        ScheduleExpression: "rate(5 minutes)"
    })
})

test('There are 11 alarms, includes 3 Latency, 3 Availability, 3 BrokenLinks', () => {
    template.resourceCountIs("AWS::CloudWatch::Alarm", 12);

    template.resourcePropertiesCountIs("AWS::CloudWatch::Alarm", {MetricName: "PageExecutionTime-ap-southeast-2-prod"}, 3)
    template.resourcePropertiesCountIs("AWS::CloudWatch::Alarm", {MetricName: "PageAvailability-ap-southeast-2-prod"}, 3)
    template.resourcePropertiesCountIs("AWS::CloudWatch::Alarm", {MetricName: "PageBrokenLinks-ap-southeast-2-prod"}, 3)
})

test('SNS has been created and has 2 email subscriptions', () => {
    template.hasResource("AWS::SNS::Topic", "");    
    template.resourceCountIs("AWS::SNS::Subscription", 2);
})

test('DynamoDB Table is created with correct properties', () => {
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        TableName: 'webcrawler-dynamodb-ap-southeast-2-prod',
        AttributeDefinitions: [
            { AttributeName: 'url', AttributeType: 'S' },
            { AttributeName: 'timestamp', AttributeType: 'S' }
        ],
        KeySchema: [
            { AttributeName: 'url', KeyType: 'HASH' },
            { AttributeName: 'timestamp', KeyType: 'RANGE' }
        ],
        ProvisionedThroughput: {
            ReadCapacityUnits: 5,
            WriteCapacityUnits: 5
        }
    })
})

test('Lambda Function for DynamoDB is created', () => {
    template.hasResourceProperties('AWS::Lambda::Function', {
        FunctionName: 'lambdadynamoDB-ap-southeast-2-prod',
        Runtime: 'nodejs20.x',
        Handler: 'dynamodb.handler'       
    })
})

test('IAM role for Lambda function is created', () => {
    template.hasResource('AWS::IAM::Role', "");
    template.hasResourceProperties('AWS::IAM::Role', {
        Description: 'Lambda IAM Role for DynamoDB-ap-southeast-2-prod'  
    })
})

test('IAM Role for Lambda interacting with DynamoDB has policies assigned', () => {
    template.hasResourceProperties('AWS::IAM::Role', {
        AssumeRolePolicyDocument: {
            Statement: [ {
                    Action: 'sts:AssumeRole',
                    Effect: 'Allow',
                    Principal: {Service: 'lambda.amazonaws.com'}}
            ]},
        Description: 'Lambda IAM Role for DynamoDB-ap-southeast-2-prod'
    })
})

test('Lambda Alias for Web Crawler is created', () => {
    template.hasResource('AWS::Lambda::Alias',"");
    template.hasResourceProperties('AWS::Lambda::Alias',{
        Name: 'prod'
    })
})

test('Lambda Deployment Group for Web Crawler is created', () => {
    template.hasResource('AWS::CodeDeploy::DeploymentGroup',"");
    template.hasResourceProperties('AWS::CodeDeploy::DeploymentGroup', {
        DeploymentConfigName: 'CodeDeployDefault.LambdaLinear10PercentEvery3Minutes',
        AutoRollbackConfiguration: {
            Events: [
                'DEPLOYMENT_FAILURE',
                'DEPLOYMENT_STOP_ON_REQUEST',
                'DEPLOYMENT_STOP_ON_ALARM'
            ]
        }
    })
})

// ======================================================== Snapshot Test =====================================================
// Might need to run "npm run test --updateSnapshot"
it('Matches the snapshot.', () => {
    expect(template.toJSON()).toMatchSnapshot();
});
    
