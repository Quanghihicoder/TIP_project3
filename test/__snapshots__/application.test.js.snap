// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches the snapshot. 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "project3webcrawlerapsoutheast2prodLambdaSourceBucketNameParameter01A90FCA": {
      "Type": "String",
    },
    "project3webcrawlerapsoutheast2prodLambdaSourceObjectKeyParameterD25890A2": {
      "Type": "String",
    },
    "webcrawlerbucketlocationParameter": {
      "Default": "webcrawler-assets-bucket-location",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "alarmhttpswwwapplecomavailabilityapsoutheast2prod1F27A44A": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for https://www.apple.com Availability Metric",
        "AlarmName": "webcrawler-alarm-https://www.apple.com-availability-ap-southeast-2-prod",
        "ComparisonOperator": "LessThanThreshold",
        "Dimensions": [
          {
            "Name": "url",
            "Value": "https://www.apple.com",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PageAvailability-ap-southeast-2-prod",
        "Namespace": "MonitorWebsites-ap-southeast-2-prod",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 1,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmhttpswwwapplecombrokenlinksapsoutheast2prodBFF4D144": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for https://www.apple.com BrokenLinks Metric",
        "AlarmName": "webcrawler-alarm-https://www.apple.com-broken-links-ap-southeast-2-prod",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "url",
            "Value": "https://www.apple.com",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PageBrokenLinks-ap-southeast-2-prod",
        "Namespace": "MonitorWebsites-ap-southeast-2-prod",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 0,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmhttpswwwapplecomlatencyapsoutheast2prodD2D5D47C": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for https://www.apple.com Latency Metric",
        "AlarmName": "webcrawler-alarm-https://www.apple.com-latency-ap-southeast-2-prod",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "url",
            "Value": "https://www.apple.com",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PageExecutionTime-ap-southeast-2-prod",
        "Namespace": "MonitorWebsites-ap-southeast-2-prod",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 3000,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmhttpswwwswinburneeduauavailabilityapsoutheast2prod958B5ED8": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for https://www.swinburne.edu.au Availability Metric",
        "AlarmName": "webcrawler-alarm-https://www.swinburne.edu.au-availability-ap-southeast-2-prod",
        "ComparisonOperator": "LessThanThreshold",
        "Dimensions": [
          {
            "Name": "url",
            "Value": "https://www.swinburne.edu.au",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PageAvailability-ap-southeast-2-prod",
        "Namespace": "MonitorWebsites-ap-southeast-2-prod",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 1,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmhttpswwwswinburneeduaubrokenlinksapsoutheast2prod8202E65F": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for https://www.swinburne.edu.au BrokenLinks Metric",
        "AlarmName": "webcrawler-alarm-https://www.swinburne.edu.au-broken-links-ap-southeast-2-prod",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "url",
            "Value": "https://www.swinburne.edu.au",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PageBrokenLinks-ap-southeast-2-prod",
        "Namespace": "MonitorWebsites-ap-southeast-2-prod",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 0,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmhttpswwwswinburneeduaulatencyapsoutheast2prod81F53CF7": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for https://www.swinburne.edu.au Latency Metric",
        "AlarmName": "webcrawler-alarm-https://www.swinburne.edu.au-latency-ap-southeast-2-prod",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "url",
            "Value": "https://www.swinburne.edu.au",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PageExecutionTime-ap-southeast-2-prod",
        "Namespace": "MonitorWebsites-ap-southeast-2-prod",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 3000,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmhttpswwwyoutubecomavailabilityapsoutheast2prod7B24955B": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for https://www.youtube.com Availability Metric",
        "AlarmName": "webcrawler-alarm-https://www.youtube.com-availability-ap-southeast-2-prod",
        "ComparisonOperator": "LessThanThreshold",
        "Dimensions": [
          {
            "Name": "url",
            "Value": "https://www.youtube.com",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PageAvailability-ap-southeast-2-prod",
        "Namespace": "MonitorWebsites-ap-southeast-2-prod",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 1,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmhttpswwwyoutubecombrokenlinksapsoutheast2prod25D46E79": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for https://www.youtube.com BrokenLinks Metric",
        "AlarmName": "webcrawler-alarm-https://www.youtube.com-broken-links-ap-southeast-2-prod",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "url",
            "Value": "https://www.youtube.com",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PageBrokenLinks-ap-southeast-2-prod",
        "Namespace": "MonitorWebsites-ap-southeast-2-prod",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 0,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmhttpswwwyoutubecomlatencyapsoutheast2prodEF8C31BB": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for https://www.youtube.com Latency Metric",
        "AlarmName": "webcrawler-alarm-https://www.youtube.com-latency-ap-southeast-2-prod",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "url",
            "Value": "https://www.youtube.com",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "PageExecutionTime-ap-southeast-2-prod",
        "Namespace": "MonitorWebsites-ap-southeast-2-prod",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 3000,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmmaxlatencyapsoutheast2prod5C330BA1": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for Max Latency Metric",
        "AlarmName": "webcrawler-alarm-max-latency-ap-southeast-2-prod",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "MAX(METRICS())",
            "Id": "expr_1",
            "Label": "Maximum Execution Time",
          },
          {
            "Id": "ml0",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "url",
                    "Value": "https://www.swinburne.edu.au",
                  },
                ],
                "MetricName": "PageExecutionTime-ap-southeast-2-prod",
                "Namespace": "MonitorWebsites-ap-southeast-2-prod",
              },
              "Period": 900,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "ml1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "url",
                    "Value": "https://www.youtube.com",
                  },
                ],
                "MetricName": "PageExecutionTime-ap-southeast-2-prod",
                "Namespace": "MonitorWebsites-ap-southeast-2-prod",
              },
              "Period": 900,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "ml2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "url",
                    "Value": "https://www.apple.com",
                  },
                ],
                "MetricName": "PageExecutionTime-ap-southeast-2-prod",
                "Namespace": "MonitorWebsites-ap-southeast-2-prod",
              },
              "Period": 900,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 3000,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "alarmminreachableapsoutheast2prodA1F1DB96": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "emailtopicapsoutheast2prod201DD58B",
          },
        ],
        "AlarmDescription": "Alarm for Min Availability Metric",
        "AlarmName": "webcrawler-alarm-min-reachable-ap-southeast-2-prod",
        "ComparisonOperator": "LessThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "SUM(METRICS())",
            "Id": "expr_1",
            "Label": "Number of availabile websites",
          },
          {
            "Id": "ma0",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "url",
                    "Value": "https://www.swinburne.edu.au",
                  },
                ],
                "MetricName": "PageAvailability-ap-southeast-2-prod",
                "Namespace": "MonitorWebsites-ap-southeast-2-prod",
              },
              "Period": 900,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "ma1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "url",
                    "Value": "https://www.youtube.com",
                  },
                ],
                "MetricName": "PageAvailability-ap-southeast-2-prod",
                "Namespace": "MonitorWebsites-ap-southeast-2-prod",
              },
              "Period": 900,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
          {
            "Id": "ma2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "url",
                    "Value": "https://www.apple.com",
                  },
                ],
                "MetricName": "PageAvailability-ap-southeast-2-prod",
                "Namespace": "MonitorWebsites-ap-southeast-2-prod",
              },
              "Period": 900,
              "Stat": "Average",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 3,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "emailtopicapsoutheast2prod201DD58B": {
      "Properties": {
        "DisplayName": "webcrawler-email-topic-ap-southeast-2-prod",
      },
      "Type": "AWS::SNS::Topic",
    },
    "emailtopicapsoutheast2prodhoangquang2508gmailcomA22A19A7": {
      "Properties": {
        "Endpoint": "hoangquang2508@gmail.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "emailtopicapsoutheast2prod201DD58B",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "project3dynamoDBlambdaapsoutheast2prodAllowInvokeProdApplicationStackemailtopicapsoutheast2prod356EC700C6A46297": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "project3dynamoDBlambdaapsoutheast2prodE2DEBD9E",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "emailtopicapsoutheast2prod201DD58B",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "project3dynamoDBlambdaapsoutheast2prodE2DEBD9E": {
      "DependsOn": [
        "project3lambdadynamodbroleapsoutheast2prodDefaultPolicyEF4E00C6",
        "project3lambdadynamodbroleapsoutheast2prod4FEDE8A4",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "project3webcrawlerapsoutheast2prodLambdaSourceBucketNameParameter01A90FCA",
          },
          "S3Key": {
            "Ref": "project3webcrawlerapsoutheast2prodLambdaSourceObjectKeyParameterD25890A2",
          },
        },
        "Environment": {
          "Variables": {
            "TABLE_NAME": {
              "Ref": "project3dynamodbapsoutheast2prod9847C4C5",
            },
          },
        },
        "FunctionName": "lambdadynamoDB-ap-southeast-2-prod",
        "Handler": "dynamodb.handler",
        "Role": {
          "Fn::GetAtt": [
            "project3lambdadynamodbroleapsoutheast2prod4FEDE8A4",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "project3dynamoDBlambdaapsoutheast2prodemailtopicapsoutheast2prod7B7FA3B3": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "project3dynamoDBlambdaapsoutheast2prodE2DEBD9E",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "emailtopicapsoutheast2prod201DD58B",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "project3dynamodbapsoutheast2prod9847C4C5": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "url",
            "AttributeType": "S",
          },
          {
            "AttributeName": "timestamp",
            "AttributeType": "S",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "url",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "timestamp",
            "KeyType": "RANGE",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "webcrawler-dynamodb-ap-southeast-2-prod",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "project3lambdadynamodbroleapsoutheast2prod4FEDE8A4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Lambda IAM Role for DynamoDB-ap-southeast-2-prod",
      },
      "Type": "AWS::IAM::Role",
    },
    "project3lambdadynamodbroleapsoutheast2prodDefaultPolicyEF4E00C6": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:GetItem",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "project3dynamodbapsoutheast2prod9847C4C5",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:CreateLogGroup",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "project3lambdadynamodbroleapsoutheast2prodDefaultPolicyEF4E00C6",
        "Roles": [
          {
            "Ref": "project3lambdadynamodbroleapsoutheast2prod4FEDE8A4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "project3webcrawleraliasapsoutheast2prodE0D51485": {
      "Properties": {
        "FunctionName": {
          "Ref": "project3webcrawlerapsoutheast2prodBDFEDA8E",
        },
        "FunctionVersion": {
          "Fn::GetAtt": [
            "project3webcrawlerapsoutheast2prodCurrentVersion546DBF54b6e63d187ab6980508f544a4bd0de2e6",
            "Version",
          ],
        },
        "Name": "webcrawleralias-ap-southeast-2-prod",
      },
      "Type": "AWS::Lambda::Alias",
      "UpdatePolicy": {
        "CodeDeployLambdaAliasUpdate": {
          "ApplicationName": {
            "Ref": "project3webcrawlerlambdaapplicationapsoutheast2prod56B124D5",
          },
          "DeploymentGroupName": {
            "Ref": "project3webcrawlerdeploymentapsoutheast2prodE5D270CA",
          },
        },
      },
    },
    "project3webcrawlerapsoutheast2prodBDFEDA8E": {
      "DependsOn": [
        "webcrawleriamroleapsoutheast2prodDefaultPolicyD418D8DA",
        "webcrawleriamroleapsoutheast2prodAA2BF0D3",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "project3webcrawlerapsoutheast2prodLambdaSourceBucketNameParameter01A90FCA",
          },
          "S3Key": {
            "Ref": "project3webcrawlerapsoutheast2prodLambdaSourceObjectKeyParameterD25890A2",
          },
        },
        "Environment": {
          "Variables": {
            "BUCKET": {
              "Ref": "webcrawlerbucketlocationParameter",
            },
            "METRIC_AVAILABILITY_NAME": "PageAvailability-ap-southeast-2-prod",
            "METRIC_BROKENLINKS_NAME": "PageBrokenLinks-ap-southeast-2-prod",
            "METRIC_LATENCY_NAME": "PageExecutionTime-ap-southeast-2-prod",
            "METRIC_NAMESPACE": "MonitorWebsites-ap-southeast-2-prod",
          },
        },
        "FunctionName": "webcrawler-ap-southeast-2-prod",
        "Handler": "webcrawler.handler",
        "LoggingConfig": {
          "LogFormat": "JSON",
          "LogGroup": {
            "Ref": "webcrawlerloggroupapsoutheast2prodD1D9A55B",
          },
        },
        "Role": {
          "Fn::GetAtt": [
            "webcrawleriamroleapsoutheast2prodAA2BF0D3",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "project3webcrawlerapsoutheast2prodCurrentVersion546DBF54b6e63d187ab6980508f544a4bd0de2e6": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "FunctionName": {
          "Ref": "project3webcrawlerapsoutheast2prodBDFEDA8E",
        },
      },
      "Type": "AWS::Lambda::Version",
      "UpdateReplacePolicy": "Delete",
    },
    "project3webcrawlerdeploymentapsoutheast2prodE5D270CA": {
      "Properties": {
        "ApplicationName": {
          "Ref": "project3webcrawlerlambdaapplicationapsoutheast2prod56B124D5",
        },
        "AutoRollbackConfiguration": {
          "Enabled": true,
          "Events": [
            "DEPLOYMENT_FAILURE",
          ],
        },
        "DeploymentConfigName": "CodeDeployDefault.LambdaCanary10Percent15Minutes",
        "DeploymentStyle": {
          "DeploymentOption": "WITH_TRAFFIC_CONTROL",
          "DeploymentType": "BLUE_GREEN",
        },
        "ServiceRoleArn": {
          "Fn::GetAtt": [
            "project3webcrawlerdeploymentapsoutheast2prodServiceRole6F423020",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CodeDeploy::DeploymentGroup",
    },
    "project3webcrawlerdeploymentapsoutheast2prodServiceRole6F423020": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codedeploy.ap-southeast-2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSCodeDeployRoleForLambdaLimited",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "project3webcrawlerlambdaapplicationapsoutheast2prod56B124D5": {
      "Properties": {
        "ApplicationName": "project3-webcrawler-lambdaapplication-ap-southeast-2-prod",
        "ComputePlatform": "Lambda",
      },
      "Type": "AWS::CodeDeploy::Application",
    },
    "webcrawleriamroleapsoutheast2prodAA2BF0D3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Web Crawler IAM Role-ap-southeast-2-prod",
      },
      "Type": "AWS::IAM::Role",
    },
    "webcrawleriamroleapsoutheast2prodDefaultPolicyD418D8DA": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:ListAllMyBuckets",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "s3:*",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:CreateLogGroup",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "webcrawleriamroleapsoutheast2prodDefaultPolicyD418D8DA",
        "Roles": [
          {
            "Ref": "webcrawleriamroleapsoutheast2prodAA2BF0D3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "webcrawlerlambdaeventapsoutheast2prod7F180FA7": {
      "Properties": {
        "Description": "Web Crawler run every 5 mins",
        "ScheduleExpression": "rate(5 minutes)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "project3webcrawlerapsoutheast2prodBDFEDA8E",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "webcrawlerlambdaeventapsoutheast2prodAllowEventRuleProdApplicationStackproject3webcrawlerapsoutheast2prod78DEB6E4F883FFF1": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "project3webcrawlerapsoutheast2prodBDFEDA8E",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "webcrawlerlambdaeventapsoutheast2prod7F180FA7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "webcrawlerloggroupapsoutheast2prodD1D9A55B": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupName": "webcrawler-log-group-ap-southeast-2-prod",
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
